test/provision_test.py

```diff
@@ -0,0 +1,108 @@
+import pytest
+import json
+
+from autoprotocol.protocol import Protocol
+
+from test.test_util import TestUtils
+
+
+class TestProvision(object):
+    p = Protocol()
+    w1 = p.ref("w1", None, cont_type="96-pcr", discard=True) \
+        .well(0).set_volume("2:microliter")
+
+    def test_provision_well_capacity(self):
+        self.p.provision("rs17gmh5wafm5p", self.w1, "50:microliter")
+        actual_protocol_as_json = json.dumps(self.p.as_dict(), indent=2, sort_keys=True)
+        expected_protocol_as_json = TestUtils.read_json_file('expected_provision.json')
```

Comment:
hmm I'm wary of this pattern. Technically, we're only modifying the provision code, so we don't want these tests to break if we were to make a separate change to a top-level key for example which is independent of the provision code.

Its debatable whether we want to use json.dumps directly, since that adds an additional fixed dependency for the serialization check. Currently we mostly check against the generated Instruction data-structure, since that's what the method directly does, and assume that the serialization portion is decoupled and works.

Reply:
Although we made changes to provision code alone, it is referring to the object `refs` in the protocol object. So, we should test provision in conjunction with `refs`. Since `refs` is a child of protocol object, I am using complete object for comparison. Moreover if we have the flexibility to create just `Provision` object alone, then I would have written test cases just around that. Since provision can be created via protocol, we should test against the entirety of protocol object.

The main intent of using json usage is to avoid boiler-plate code for assertions. There are no custom serialization in any of these methods.

Reply:
had a discussion offline about this. 
In summary, this is a viable approach if we were to just test against Instructions. It forces the developer to update each json file if a new Instruction field is added, which trades off test maintenance with potentially edge-case discovery. The hope there is that by forcing the developer to update each file, he may discover some dependent relationship. Its worth trialing this for this Instruction.